packages/@aws-cdk/aws-dynamodb/lib/table-time-to-live-handler/index.ts

```diff
@@ -0,0 +1,66 @@
+/* eslint-disable no-console */
+import type { IsCompleteRequest, IsCompleteResponse, OnEventRequest, OnEventResponse } from '@aws-cdk/custom-resources/lib/provider-framework/types';
+import { DynamoDB } from 'aws-sdk'; // eslint-disable-line import/no-extraneous-dependencies
+import { disableTimeToLive, enableTimeToLive, isTimeToLiveCorrect, isTimeToLiveStable } from './utils';
+
+const dynamodb = new DynamoDB({ apiVersion: '2012-08-10' });
+
+export async function onEventHandler(event: OnEventRequest): Promise<OnEventResponse> {
+  console.log('Event: %j', event);
+
+  if (event.RequestType === 'Create' || event.RequestType === 'Update') {
+    const currentTtl = await dynamodb.describeTimeToLive({
+      TableName: event.ResourceProperties.TableName,
+    }).promise();
+    console.log('Describe time to live: %j', currentTtl);
+
+    switch (currentTtl.TimeToLiveDescription?.TimeToLiveStatus) {
+      case 'ENABLING':
+        throw new Error('Can not change time to live while it is being enabled.');
+        break;
+      case 'DISABLING':
+        throw new Error('Can not change time to live while it is being disabled.');
+        break;
+      case 'ENABLED':
+        if (event.ResourceProperties.TimeToLiveAttribute !== currentTtl.TimeToLiveDescription.AttributeName) {
+          await disableTimeToLive(event);
+        }
+        break;
+      case 'DISABLED':
+        if (event.ResourceProperties.TimeToLiveAttribute) {
+          await enableTimeToLive(event);
+        }
+        break;
+    }
+
+    console.log('Update table: %j', event.ResourceProperties.TableName);
+    return { PhysicalResourceId: event.PhysicalResourceId ?? event.LogicalResourceId };
+  }
+
+  return { PhysicalResourceId: event.PhysicalResourceId };
+}
+
+export async function isCompleteHandler(event: IsCompleteRequest): Promise<IsCompleteResponse> {
+  console.log('Event: %j', event);
+
+  const data = await dynamodb.describeTable({
+    TableName: event.ResourceProperties.TableName,
+  }).promise();
+  console.log('Describe table: %j', data);
+
+  const tableActive = (data.Table?.TableStatus === 'ACTIVE');
+
+  if (await isTimeToLiveStable(event)) {
+    await onEventHandler(event);
+  }
+
+  switch (event.RequestType) {
+    case 'Create':
+    case 'Update':
+      // Complete when time to live is fully functional
+      return { IsComplete: tableActive && await isTimeToLiveStable(event) && await isTimeToLiveCorrect(event) };
```

Comment:
How could it not be correct?

Reply:
As mentioned above, if we want to change the TTL Attribute from MyTTL to MyNewTTL, we first need to disable it. If this process completes the TTL will be “Stable” but it will be inactive, whereas the template asked for a new active ttl. In most cases the above call to `onEventHandler` will result in the ttl becoming _“instable”_ again, but there could _potentially_ be a race condition where the above returns false but here it returns true.

Reply:
I think we need to expand this `switch` case, to put the code above calling `isTimeToLiveStable` here.

I think we're doing a lot of duplicate calls to `describeTimeToLive()`, and I think we can simplify this logic quite a bit if we save the results of doing a single call, and then read them here, in this `switch` branch.

Reply:
I agree that we can probably reduce the number of API calls, but we will need to do it at least two times, as the result can change after the call of `onEventHandler(...)`. I have updated the code to only do a second API call when `onEventHandler(...)` is called as the result doesn’t change otherwise. Let me know what you think of the changes.